feat(disk_io): added disk I/O module

vilkoz opened the pull request:
Added disk IO speed module
At the moment the module shows sum of read / write speds of all block devices.
Improvement suggestions are welcome.

NBonaparte commented on doc/config.cmake:
```diff
@@ -394,6 +394,14 @@ menu-2-0-exec = sudo poweroff
+format-prefix = " "
+format-prefix-foreground = ${colors.foreground-alt}
+format-underline = #4bffdc
+label = R: %speed_read% Mb/s W: %speed_write% Mb/s
```
The units should be part of `%speed_read%` and `%speed_write%`.

NBonaparte commented on src/modules/disk_io.cpp:
```diff
@@ -0,0 +1,157 @@
+  return time_delta;
+}
+
+void disk_io_module::_calculate_disk_io_speeds(std::string disk_name)
```
You should use `string_util::filesize()` if you wanted (kilo/mega)bytes/s, or create another `string_util` function to calculate it in (kilo/mega)bits/s.

NBonaparte commented on include/modules/disk_io.hpp:
```diff
@@ -0,0 +1,44 @@
+      bool build(builder* builder, const string& tag) const;
+
+    private:
+      std::vector<std::string> _get_disk_names(void);
```
No need to start helper functions with underscores.

NBonaparte commented on src/modules/disk_io.cpp:
```diff
@@ -0,0 +1,157 @@
+    std::getline(disk_io_info, line);
+
+    static const std::regex re{"\\s+"};
+    std::vector<std::string> items{
```
For easier parsing (without variable length whitespaces) you could use `/proc/diskstats/` and split it into a vector with `string_util::split()`.

vilkoz replied:
the `/proc/diskstats` actually also has continuous sequence of spaces so I can't easily split it to needed values without regular expression

NBonaparte commented on src/modules/disk_io.cpp:
```diff
@@ -0,0 +1,157 @@
+    }
+    m_disk_names = _get_disk_names();
+    for (auto name : m_disk_names) {
+      std::cout << "name: " << name << std::endl;
```
Not needed, unless you want to use `m_log.info()` to output the names.

vilkoz commented:
This module allows user to select which disk to monitor, if multiple selected - it will show sum of speeds, to minimize consumption of available space on bar.

NBonaparte commented:
Are you planning to implement `bar-read` and `bar-write`, or would you like to merge this as-is?

vilkoz commented:
I've planned to add them as soon as I have a free minute

vilkoz commented:
FYI: I've added bar-read and bar-write, but they don't seem to be very usefull in context of disk read/write speeds. BTW, I think this feature is ready to be merged.

patrick96 commented on src/modules/disk_io.cpp:
```diff
@@ -0,0 +1,169 @@
+  }
+
+  m_max_read_speed = std::max(m_max_read_speed, sum_read);
+  m_max_write_speed = std::max(m_max_write_speed, sum_write);
```
Are the max speeds just the highest speeds measured during the time the module was running?

vilkoz replied:
Yes, this method will determine maximum disk read/write speed as soon as read/write to disk will continue longer then module update period. Of course you can perform test write and read from disk for determining this values at initialization of the module, but for me it seems to be a little owerkill for module that will run at long period of time

patrick96 replied:
So, if any of the disks don't reach their maximum speeds then the values here are going to be inaccurate?

vilkoz replied:
The situation where disks don't reach max speed is highly unlikely and hard to reproduce. But those calculations only made for `<bar-write>` and `<bar-read>`, which are not seem to be very informative at all for displaying precise numbers, they are more suitable for displaying changes in I/O speed, e.g. when you start download directly to hdd, you will see that bar changed dramatically from 0% to 100%, and when download is finished - effect will be opposite.
I've got an idea when I wrote the last sentence: old computers have LED indicators for disk usage, and they (indicators) are quite informative, so you can easily detect if your machine performs some I/O intense actions. Maybe it would be more useful to implement analog for such indicator, using ramp drawtype, as a replacement for bar-read bar-write?

vilkoz commented:
Hello, I have replaced bar-read/write with indicator-read/write, they informative to the same extent but take much less space on bar. Also, I've added some explanatory comments to the module in config, so configuring of it can be done without exploring the source code.
In my opinion the module is ready to be merged.

jvandenbroek commented:
Any change this gets merged?

vilkoz commented:
@jvandenbroek They seems to make plugin architecture for providing additional features, so now you can just use bash script to replace this feature

vilkoz commented:
@NBonaparte does this pull-request needs some changes, or it should be abandoned?

ironyman commented:
@vilkoz can we address the late emperor's comments and get this feature merged? :)

patrick96 commented:
This is on my TODO list to be reviewed. I don't have a lot of free time and there are a bunch of higher priority issues, so that's why this hasn't been reviewed yet.

lockie commented:
Any chance of adding this to `master`?

derrgill commented:
Really looking forward to a disk I/O module, to have some semblance of parity with the Gnome System Monitor applet / extension that's been around in some form since the Gnome 2.x days
